src/displayapp/screens/Calculator.cpp

```diff
@@ -0,0 +1,350 @@
+      return value;
+    };
+  };
+  
+  
+  struct BinOp : CalcTreeNode {
+    CalcTreeNode* left;
+    CalcTreeNode* right;
+    
+    char op;
+    
+    double calculate() override {
+      switch (op) {
+        case '^':
+          return pow(left->calculate(), right->calculate());
+        case 'x':
+          return left->calculate() * right->calculate();
+        case '/':
+          return left->calculate() / right->calculate();
+        case '+':
+          return left->calculate() + right->calculate();
+        case '-':
+          return left->calculate() - right->calculate();
+      }
+      return 0;
+    };
+    
+  };
+  
+  uint8_t getPrecedence(char op) {
+    switch (op) {
+      case '^':
+        return 4;
+      case 'x':
+      case '/':
+        return 3;
+      case '+':
+      case '-':
+        return 2;
+    }
+    return 0;
+  }
+  
+  bool leftAssociative(char op) {
+    switch (op) {
+      case '^':
+        return false;
+      case 'x':
+      case '/':
+      case '+':
+      case '-':
+        return true;
+    }
+    return false;
+  }
+  
+}
+
+
+static void eventHandler(lv_obj_t* obj, lv_event_t event) {
+  auto calc = static_cast<Calculator*>(obj->user_data);
+  calc->OnButtonEvent(obj, event);
+}
+
+Calculator::~Calculator() {
+  lv_obj_clean(lv_scr_act());
+}
+
+static const char* buttonMap1[] = {"7", "8", "9", "/", "\n",
+                                   "4", "5", "6", "x", "\n",
+                                   "1", "2", "3", "-", "\n",
+                                   ".", "0", "=", "+", "",};
+
+static const char* buttonMap2[] = {"7", "8", "9", "(", "\n",
+                                   "4", "5", "6", ")", "\n",
+                                   "1", "2", "3", "^", "\n",
+                                   ".", "0", "=", "+", "",};
+
+Calculator::Calculator(DisplayApp* app, Controllers::MotorController& motorController)
+    : Screen(app), motorController{motorController} {
+  result = lv_label_create(lv_scr_act(), nullptr);
+  lv_label_set_long_mode(result, LV_LABEL_LONG_BREAK);
+  lv_label_set_text(result, "0");
+  lv_obj_set_size(result, 180, 60);
+  lv_obj_set_pos(result, 0, 0);
+  
+  returnButton = lv_btn_create(lv_scr_act(), nullptr);
+  lv_obj_set_size(returnButton, 52, 52);
+  lv_obj_set_pos(returnButton, 180, 8);
+  lv_obj_t* returnLabel;
+  returnLabel = lv_label_create(returnButton, nullptr);
+  lv_label_set_text(returnLabel, "<=");
+  lv_obj_align(returnLabel, nullptr, LV_ALIGN_CENTER, 0, 0);
+  returnButton->user_data = this;
+  lv_obj_set_event_cb(returnButton, eventHandler);
+  
+  buttonMatrix = lv_btnmatrix_create(lv_scr_act(), nullptr);
+  lv_btnmatrix_set_map(buttonMatrix, buttonMap1);
+  lv_obj_set_size(buttonMatrix, 240, 180);
+  lv_obj_set_pos(buttonMatrix, 0, 60);
+  buttonMatrix->user_data = this;
+  lv_obj_set_event_cb(buttonMatrix, eventHandler);
+  running = true;
+}
+
+void Calculator::eval() {
+  std::stack<char> input(std::deque<char>(0, ' '));
```

Comment:
I don't see any tangible drawbacks from doing it this way.

Reply:
I agree the differences are pretty negligible. The main difference I could see is that iterating backwards by index instead of forwards is less intuitive and slightly more bugprone. Eg. If you iterate backwards and accidentally used an unsigned index variable instead of a signed one your loop never exits. There's nothing incorrect in your code that I can see, it's just something that I noticed.

Reply:
it doesn't really make a difference either way and i like the stack. I'll leave it like that